plugins/chat/lib/chat_channel_hashtag_data_source.rb

```diff
@@ -44,4 +44,31 @@ def self.search(guardian, term, limit)
   def self.search_sort(search_results, _)
     search_results.sort_by { |result| result.text.downcase }
   end
+
+  def self.search_without_term(guardian, limit)
+    if SiteSetting.enable_experimental_hashtag_autocomplete
+      allowed_channel_ids_sql =
+        Chat::ChatChannelFetcher.generate_allowed_channel_ids_sql(
+          guardian,
+          exclude_dm_channels: true,
+        )
+      channel_ids = DB.query(<<~SQL, limit: limit).map(&:chat_channel_id)
+        SELECT chat_channel_id, MAX(chat_messages.created_at)
+        FROM chat_messages
+        INNER JOIN chat_channels ON chat_messages.chat_channel_id = chat_channels.id
+        WHERE chat_messages.deleted_at IS NULL
+          AND chat_channels.deleted_at IS NULL
+          AND chat_channel_id IN (#{allowed_channel_ids_sql})
+          AND chat_messages.created_at > (NOW() - INTERVAL '2 WEEKS')
+        GROUP BY chat_channel_id
+        ORDER BY MAX(chat_messages.created_at) DESC
+        LIMIT :limit
```

Comment:
I think we'll need to store a cache on `chat_channel` with regards to the last created chat message. Otherwise, this is going to be expensive with the number of rows we need to process every single time there is an empty search.

Reply:
Ah so a `latest_chat_message_id` on the channel similar to `latest_topic_id` on category?